tailfs/tailfsimpl/dirfs/dirfs_test.go

```diff
@@ -0,0 +1,353 @@
+	fs, _, _, _, close := createFileSystem(t)
+	defer close()
+
+	tests := []struct {
+		label string
+		name  string
+		err   error
+	}{
+		{
+			label: "attempt to remove root folder",
+			name:  "/",
+			err:   os.ErrPermission,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.RemoveAll(ctx, test.name)
+			if err == nil || !errors.Is(err, test.err) {
+				t.Errorf("expected error: %v   got: %v", test.err, err)
+			}
+		})
+	}
+}
+
+func TestRename(t *testing.T) {
+	fs, _, _, _, close := createFileSystem(t)
+	defer close()
+
+	tests := []struct {
+		label   string
+		oldName string
+		newName string
+		err     error
+	}{
+		{
+			label:   "attempt to move root folder",
+			oldName: "/",
+			newName: "/domain/remote2/copy.txt",
+			err:     os.ErrPermission,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.Rename(ctx, test.oldName, test.newName)
+			if err == nil || test.err.Error() != err.Error() {
+				t.Errorf("expected error: %v   got: %v", test.err, err)
+			}
+		})
+	}
+}
+
+func createFileSystem(t *testing.T) (webdav.FileSystem, string, string, *tstest.Clock, func()) {
+	l1, dir1 := startRemote(t)
+	l2, dir2 := startRemote(t)
+
+	// Make some files, use perms 0666 as lowest common denominator that works
+	// on both UNIX and Windows.
+	err := os.WriteFile(filepath.Join(dir1, "file1.txt"), []byte("12345"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.WriteFile(filepath.Join(dir2, "file2.txt"), []byte("54321"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// make some directories
+	err = os.Mkdir(filepath.Join(dir1, "dir1"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.Mkdir(filepath.Join(dir2, "dir2"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clock := tstest.NewClock(tstest.ClockOpts{Start: time.Now()})
+	fs := &FS{
+		Clock:      clock,
+		StaticRoot: "domain",
+		Children: []*Child{
+			{Name: "remote1"},
+			{Name: "remote2"},
+			{Name: "remote4"},
+		},
+	}
+
+	return fs, dir1, dir2, clock, func() {
+		defer l1.Close()
+		defer os.RemoveAll(dir1)
+		defer l2.Close()
+		defer os.RemoveAll(dir2)
+	}
+}
+
+func startRemote(t *testing.T) (net.Listener, string) {
+	dir := t.TempDir()
+
+	l, err := net.Listen("tcp", "127.0.0.1:")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	h := &webdav.Handler{
+		FileSystem: webdav.Dir(dir),
+		LockSystem: webdav.NewMemLS(),
+	}
+
+	s := &http.Server{Handler: h}
+	go s.Serve(l)
+
+	return l, dir
+}
+
+func infosEqual(t *testing.T, expected, actual fs.FileInfo) {
+	t.Helper()
+	if expected.Name() != actual.Name() {
+		t.Errorf("expected name: %v   got: %v", expected.Name(), actual.Name())
+	}
+	if expected.Size() != actual.Size() {
+		t.Errorf("expected Size: %v   got: %v", expected.Size(), actual.Size())
+	}
+	if !expected.ModTime().Truncate(time.Second).UTC().Equal(actual.ModTime().Truncate(time.Second).UTC()) {
```

Comment:
Hmm, shouldn't the times be able to be `expected.ModTime().Equal(actual.ModTime())` exactly? We control the clock on the implementation AFAIK.

Reply:
You are correct, this was vestigial from when we had a webdav.Handler in the way that truncated times to meet the WebDAV spec.